src/EditorFeatures/Core.Wpf/PasteActions/SuggestedPasteActionsSource.cs

```diff
@@ -0,0 +1,370 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editor.Implementation.Suggestions;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PasteActions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.PasteActions
+{
+    internal partial class SuggestedPasteActionsSourceProvider : ISuggestedActionsSourceProvider
+    {
+        internal class SuggestedPasteActionsSource : ForegroundThreadAffinitizedObject, ISuggestedActionsSource2
+        {
+            private readonly ISuggestedActionCategoryRegistryService _suggestedActionCategoryRegistry;
+
+            // state that will be only reset when source is disposed.
+            private SuggestedPasteActionsSourceProvider _owner;
+            private ITextView _textView;
+            private ITextBuffer _subjectBuffer;
+            private WorkspaceRegistration _registration;
```

Comment:
there seems to be a lot of duplication between this Source, and our regular Source.  Can we not share a lot of code between the two, and only introduce specialized behavior handling where necessary?